### PR TITLE
Emailing

### PR DESCRIPTION
**Pull request by InouRidder**
I've branched off https://github.com/TableVibe/tablevibes-backend/pull/39 
Please merge that first. 

DONE: 
  Add reset password flow
  Add sendgrid details for production

### CODE REVIEW

File: config/initializers/smtp.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+# Setting base smtp settings to configure sendgrid in production
+ActionMailer::Base.smtp_settings = {
+  user_name: Rails.application.credentials.sendgrid_user_name,
```
Comment (DanielGordon1) [inline config/initializers/smtp.rb]: ... Please explain how this is nicer than ENV['SENDGRID_USER_NAME']

File: app/models/user.rb [hunk 2]
```diff
@@ -22,7 +24,7 @@ class User < ApplicationRecord
            :employees, :reviews, :vouchers,
            to: :account
 
-  validates :first_name, :last_name, presence: true, on: :update
+  # validates :first_name, :last_name, presence: true, on: :update, unless: :invited
```
Comment (DanielGordon1) [inline app/models/user.rb]: This seems like something we should uncomment

File: app/views/devise/mailer/email_changed.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
```
Comment (DanielGordon1) [inline app/views/devise/mailer/email_changed.html.erb]: &. > try

File: app/views/user_mailer/send_invitation.html.erb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    <p>
+      Click the button below to set your password
+    </p>
+    <div class="spacer" style="width: 100%;height: 5px;"></div>
+    <div class="spacer" style="width: 100%;height: 5px;"></div>
+    <p>
+      Tablevibe
+    </p>
+  </div>
+  <a href='<%= "#{Rails.application.credentials.send("client_ui_url_#{Rails.env}")}/login?token=#{@token}" %> ' class="button" style="text-decoration: none;">
```
[the remark this answers is not in the document]
Comment (DanielGordon1) [reply app/views/user_mailer/send_invitation.html.erb]: Aside from the 'it sucks' part, what does send actually do?

File: config/application.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@
 require "action_mailer/railtie"
 require "action_view/railtie"
 # require "action_cable/engine"
-# require "sprockets/railtie"
+require "sprockets/railtie" if Rails.env.development?
```
Comment (DanielGordon1) [inline config/application.rb]: Why is this needed?

File: spec/mailers/previews/user_voucher_preview.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_voucher
+class UserVoucherPreview < ActionMailer::Preview
```
Comment (DanielGordon1) [inline spec/mailers/previews/user_voucher_preview.rb]: Could you rename this class?

File: spec/models/review_spec.rb [hunk 7]
```diff
@@ -66,4 +66,20 @@
...
     end
   end
+
+  describe '#call_user_voucher_mailer :on_create callback' do
+    let(:venue) { build(:venue) }
+
+    it 'should send an email when a review is created' do
+      review = create(:review, venue: venue, skip_sending_email: false)
+
+      expect(ActionMailer::Base.deliveries.map(&:to).flatten).to include(review.reviewer_email)
```
Comment (DanielGordon1) [inline spec/models/review_spec.rb]: Why did you change this?

The previous test was working fine afaik

File: Gemfile [hunk 8]
```diff
@@ -42,13 +42,16 @@ group :development, :test do
...
-  gem 'pry-rails'
+  gem 'factory_bot_rails'
   gem 'faker'
+  gem 'pry-rails'
   gem 'rspec-rails'
-  gem 'factory_bot_rails'
 end
 
 group :development do
+  gem 'launchy'
```
Comment (DanielGordon1) [inline Gemfile]: Does letter_opener_web require these gems?

Comment (DanielGordon1) [reply Gemfile]: If not then lets get rid of them :)

File: config/initializers/smtp.rb (hunk 1 above)
Comment (InouRidder) [reply config/initializers/smtp.rb]: Because if you use credentials you do not need to upload any key to any environment, not share it between developers nor think about it when deploying. You upload once the master key to decode the credentials and thats it. So instead of having to worry about giving each other the keys, we throw whatever we need in the credentials and it will be available everywhere. 
I think that definitely beats ENV for me.

File: app/views/user_mailer/send_invitation.html.erb (hunk 4 above)
[the remark this answers is not in the document]
Comment (InouRidder) [reply app/views/user_mailer/send_invitation.html.erb]: send allows you to send a dynamically named method 

`object.say_hi === object.send("say_hi")`
So it allows you to dynamically change the method you are calling, in this case I've added the urls in the credentials with production/development appended. So this way this email works in prod as well as dev.

[the remark this answers is not in the document]
Comment (InouRidder) [reply app/views/user_mailer/send_invitation.html.erb]: I don't see why it sucks except for your personal vendetta vs `Rails.credentials` 😂

File: app/models/user.rb (hunk 2 above)
Comment (InouRidder) [reply app/models/user.rb]: There is not a requirement anywhere for it to be validated and it was adding difficulty to the `invite user to your account` flow. So leaving it uncommented until clients specifically say this is a necessity.

File: spec/models/review_spec.rb (hunk 7 above)
Comment (InouRidder) [reply spec/models/review_spec.rb]: It wasn't because > it only worked when there was a single email callback throughout the entire structure. But assuming there will be more, just assuming that the review count does not increase becomes a bit unreliable imo

File: config/application.rb (hunk 5 above)
Comment (InouRidder) [reply config/application.rb]: letter_opener_web needs assets : ( bit of a shit dependency, this way at least we are only loading sprockets in dev when we actually use letter_opener_web

File: app/views/devise/mailer/email_changed.html.erb (hunk 3 above)
Comment (InouRidder) [reply app/views/devise/mailer/email_changed.html.erb]: this is the default generated by devise. I didn't touch this 🌮

File: spec/mailers/previews/user_voucher_preview.rb (hunk 6 above)
Comment (InouRidder) [reply spec/mailers/previews/user_voucher_preview.rb]: fixed

File: Gemfile (hunk 8 above)
Comment (InouRidder) [reply Gemfile]: they be gone

File: spec/models/review_spec.rb (hunk 7 above)
Comment (DanielGordon1) [reply spec/models/review_spec.rb]: Fair point! 

But we should be able to test if this specific method just sends 1 email without other callbacks being fired, perhaps we can look at disabling the other email callbacks in certain cases.

File: config/initializers/smtp.rb (hunk 1 above)
Comment (InouRidder) [reply config/initializers/smtp.rb]: yea I guess syntactically its not much of an upgrade. But we don't have to debug production or a collague's localhost because of any missing keys, which me likey. I guess its a trade off.

File: app/models/user.rb (hunk 2 above)
Comment (InouRidder) [reply app/models/user.rb]: removed 🌮

File: spec/models/review_spec.rb (hunk 7 above)
Comment (InouRidder) [reply spec/models/review_spec.rb]: jep, that would be great, I have had this email test struggle before since after a while a lot emails get triggered and hard to define what is what - so far I end up testing that the array includes the email I want to send, which isn't a great isolated test.

**Comment by InouRidder** [comment]
merging
